File: Algorithms/1D/Binary_Search/Binary_Search.cc
#include "../../../HEADER.h"

int main() {
  i64 n;
  cin >> n;

  vi64 arr(n);
  for (i64 i : range(n)) {
    cin >> arr[i];
  }

  // vector needs to be sorted before binary search

  srt(arr);

  i64 l = 0;
  i64 r = n - 1;
  i64 x = 3; // we search for x

  while (l <= r) {
    i64 mid = l + (r - l) / 2;

    if (arr[mid] == x) {
      cout << "Found" << endl;
      break;
    }

    if (arr[mid] < x) {
      l = mid + 1;
    }

    if (arr[mid] > x) {
      r = mid - 1;
    }
  }

  return 0;
}

// time complexity O(log size(arr))

File: Algorithms/1D/Binary_Search/Binary_Search.py
def main():
    arr = [2, 5, 3, 20, 7]

    arr.sort()  # the list needs to be sorted

    l = 0
    r = len(arr) - 1
    x = 2

    while l <= r:
        mid = l + (l - r) / 2

        if arr[mid] == x:
            print("FOUND")
            break

        if arr[mid] < x:
            l = mid + 1

        if arr[mid] > x:
            r = mid - 1


main()

# Time complexity O(log size(arr))
